app/services/fixity_checker.rb

```diff
@@ -0,0 +1,117 @@
+# FixityChecker.new(asset).check
+# FixityChecker.new(asset).prune_checks
+# This first version of the fixity checker only checks
+# for actual discrepancies between already-recorded
+# checksums and calculated ones.
+#
+# This means:
+# Assets with nil files are ignored.
+# Files with nil checksums are ignored.
+
+class FixityChecker
+  attr_reader :asset
+
+  NUMBER_OF_RECENT_PASSED_CHECKS_TO_KEEP = 5
+  CHECK_CYCLE_LENGTH = 7
+
+
+  # @param work [Work] Work object, it's members will be put into a zip
+  # @param callback [proc], proc taking keyword arguments progress_i: and progress_total:, can
+  #   be used to update a progress UI.
+  def initialize(asset)
+    raise ArgumentError.new("Please pass in an Asset.") unless asset.is_a? Asset
+    @asset = asset
+  end
+
+  #FixityChecker.new(asset).check
+  def check(sieve_integer = 0)
+    return unless sift(sieve_integer)
+    return nil if @asset.nil?
+    return nil if @asset.file.nil?
+    the_expected_checksum = expected_checksum
+    # Note: @asset.file.sha512 can be nil on e.g.
+    # items imported with DISABLE_BYTESTREAM_IMPORT=true
+    # For now, we are ignoring these.
+    return nil if the_expected_checksum.nil?
+
+    new_check = FixityCheck.new(asset: @asset)
+    new_check.expected_result = the_expected_checksum
+    new_check.checked_uri = @asset.file.url
+    new_check.actual_result = actual_checksum
+    new_check.passed= (the_expected_checksum==actual_checksum)
+    new_check.save!
```

Comment:
Let's do this all in one statement like this:

```ruby
FixityCheck.create!(
  asset: @asset,
  checked_uri: @asset.file.url,
  actual_result: actual_checksum,
  passed: (the_expected_checksum==actual_checksum)
)
```

But that makes me think of a couple things. 

1. Should we be storing the expected checksum too in the FixityCheck log, as an added check, in case it gets corrupted or disappears in the 'main' place it is stored? So we have a record of exactly what we expected -- in case of passing, that's the same as "actual", but in case of failing it might be different, and we might want it recorded here. 

2. Saving @asset.file.url is a bit odd. For instance, if it's on S3 and shrine doesn't think it's a public ACL, it'll be a time-limited signed S3 URL which won't actually be accessible in the far future. So saving URL isn't quite right. We could try to come up with the 'right' thing to save... but I wonder if we need it at all, the asset ID (along with _expected_ as well as actual checksum) might be enough to preserve?

3. As discussed, let's store that the checksums are SHA512 somewhere. Another column, or a prefix "sha512:[checksum]", not sure which is best, another column is fine if you think so. 

Also, how about calling the method `expected_checksum` to match `actual_checksum` instead of `the_expected_checksum`, `the` in method name is unnecessary and mismatched. 
)

And let's make the method names match the column/attribute names in the FixityCheck model. Either `actual_checksum` for both or `actual_result` for both, not `checksum` for one and `result` for the other.

Reply:
If you do want to get a reliable non-time-limited non-signed URL from S3, you can call `@asset.file.url(public: true)` to ensure it. 

It may give you an access denied if you actually try to access it and it wasn't public, but it will identify the file location in a clear way.

Shrine storage locations that aren't S3 will just ignore the `public:true`... I think. Unless they raise on it. 

Shrine's back-end storage agnosticism ends up making it a bit difficult to get the "really, where was this thing actually stored" in a clear way. Not totally sure if we need to try to store that or not.

Reply:
Using the public:true -- I definitely don't want these URIs expiring on us. The idea is to make sure everything works in case an asset gets a different file some day. (Kind of an edge case.)